lld: Add initial support for GNU LTO format

tuliom opened the pull request:
This is not feature complete yet, but is able to link simple executables.

The feature is disabled by default and can be enable via
`-DLLD_LINK_GPL3:BOOL=ON -DLLD_ENABLE_GNU_LTO:BOOL=ON`.

This implementation reuses GCC's plugin `liblto_plugin.so`, loading it at
runtime, when necessary.

This proposal has been discussed in the RFC:
https://discourse.llvm.org/t/rfc-lld-add-support-for-gcc-lto-format/87172

tuliom committed:
lld: Add support for -plugin

Modify current tests in order to support -plugin.

tuliom committed:
lld: Split the LTO code in order to add support for GCC LTO

Rename classes BitcodeFile and BitcodeCompiler to IRFile and IRCompiler
respectively.  This helps to reuse their code in the new GCC-related
classes while keeping their names semantically correct.

Part of the code from IRCompiler that is specific to LLVM got moved to a
new class called BitcodeCompiler.
The new class GccIRCompiler abstracts the interface to GCC's
liblto_plugin.so.

tuliom committed:
lld: Add initial code for liblto_plugin

Implement the most basic functions from GccIRCompiler that let LLD link
simple executables.

Add a config.h file in order to detect at configure time if
liblto_plugin supports version 2 of LDPT_REGISTER_CLAIM_FILE_HOOK.

tuliom committed:
lld: Implement options to disable GNU LTO and GPLv3 linking

Add LLD_LINK_GPL3 and LLD_ENABLE_GNULTO in order to control if LLD can
link/load to GPLv3 code and enabling support for the GNU LTO format
supported by GCC.

tuliom committed:
lld: Add a test for GNU LTO

This test ensures that GCC and LLD are both used and the final binary
has GNU LTO.

tuliom committed:
Fix code formatting issues

tuliom committed:
Fix a code formatting issue

tuliom committed:
Fix the build when plugin-api.h is not available

tuliom committed:
Fix warnings reported by the CI

tuliom committed:
Fix the build when dlfcn.h is not available

e.g. on Windows.

tuliom committed:
Merge branch 'main' into lld-plugin

tuliom committed:
Improve the description of LLD_LINK_GPL3

Specify liblto_plugin.so and try to distinguish from the usual linking
that happens when linking to libstdc++ or libgcc_s.

tuliom commented on lld/CMakeLists.txt:
```diff
@@ -177,6 +177,14 @@ if (LLD_DEFAULT_LD_LLD_IS_MINGW)
   add_definitions("-DLLD_DEFAULT_LD_LLD_IS_MINGW=1")
 endif()
 
+option(LLD_LINK_GPL3 "Allow LLD to link to GPLv3-licensed files")
```
In my last update I specified the GCC compiler plugin and tried to restrict the explanation to plugins.

smithp35 reviewed (commented):
I've left some quick comments, I'll need to spend some time looking at how the plugin works. 

My apologies, I'll be out of the office for the next 2 weeks and then at the US developer meeting. I can go through this is in detail when I get back. I don't have any particularly strong opinions so don't let this block anyone else.

Before I, or anyone else spends a lot of time, I'd like to know if there is anyone with any fundamental objections. I know in the past that there has been some reluctance (https://github.com/llvm/llvm-project/issues/41791) but my reading of the RFC https://discourse.llvm.org/t/rfc-lld-add-support-for-gcc-lto-format/87172/1 looks like positions have moved on since then. 

With the obligatory "I am not a lawyer but" disclaimer, I do think there is a material difference between `#include <plugin-api.h>`, the GNU lto plugin and linking against libgcc and libstdc++ as the LTO includes and library don't have the GCC RUNTIME LIBRARY EXCEPTION in their license. As I understand it from the RFC the foundations lawyers are OK with this (perhaps worth citing the RFC in the description).

tuliom committed:
Replace the usage of dlopen with llvm::sys::DynamicLibrary

This will be helpful in order to support environments where dlfcn.h is
not available, e.g. Windows + mingw.

tuliom committed:
Remove unnecessary tests for -plugin

tuliom committed:
Fix code format

tuliom committed:
Remove code execution from hello.test

This could cause crashes on cross compilation environments.
Replace it with a check if the main function is indeed listed.

tstellar commented on lld/ELF/LTO.h:
```diff
@@ -29,32 +30,96 @@
 };
+
+#if LLD_ENABLE_GNU_LTO
+#include <plugin-api.h>
```
Ok, make sense.  I'm fine copying it if it helps reduce dependencies.

tuliom replied:
I've re-implemented the API in C++.
I believe this point has been resolved.

tuliom committed:
Replace PATH manipulation on tests with -B

GCC provides parameter -B in order to achieve the same results.

tuliom committed:
Replace raw pointer usage with SmallVector

tuliom committed:
Remove wrappers regClaimFile and regClaimFileV2

Peter0x44 commented:
Thanks for doing this. I've wanted a faster replacement of ld.bfd for windows, supporting GCC lto, and this could be it.

tuliom committed:
Replace error() with Err(ctx)

tuliom committed:
Remove unnecessary semicolons

tuliom committed:
Reuse code from compileBitcodeFiles in compileGCCIRFiles

Create a new function called compileFiles that can be reused both in
compileBitcodeFiles() and compileGCCIRFiles().

tuliom committed:
Replace the singleton with a global variable

tuliom committed:
Add GnuLTO.h as a replacement for plugin-api.h

Rewrite the API used by GCC in C++ allowing LLD to be built without
depending on plugin-api.h being installed on the system, which
simplifies the bootstrap of a toolchain.

tuliom committed:
Merge branch 'main' into lld-plugin

tuliom committed:
Fix coding style issues

tuliom committed:
Remove config.operating_system

This is not used anymore.

tuliom commented on lld/test/ELF/gnu-lto/hello.test:
```diff
@@ -0,0 +1,28 @@
+// REQUIRES: gnu_lto, gcc
```
@smithp35 , @MaskRay the tests are not trying to execute the linked executables anymore.

tuliom commented on lld/test/lit.cfg.py:
```diff
@@ -182,3 +182,25 @@
     config.excludes.append("ELF")
+
+
+def prepend_path(path):
```
Done.

tuliom commented on lld/ELF/LTO.cpp:
```diff
@@ -421,3 +420,242 @@ SmallVector<std::unique_ptr<InputFile>, 0> BitcodeCompiler::compile() {
+
+  // TODO: Properly find the right size.
+  int tvsz = 100;
+  tv = new ld_plugin_tv[tvsz];
```
Done.

tuliom commented on lld/ELF/LTO.cpp:
```diff
@@ -421,3 +420,242 @@ SmallVector<std::unique_ptr<InputFile>, 0> BitcodeCompiler::compile() {
+  (*onload)(tv);
+}
+
+enum ld_plugin_status regClaimFile(ld_plugin_claim_file_handler handler) {
```
Done.

tuliom commented on lld/ELF/LTO.cpp:
```diff
@@ -421,3 +420,242 @@ SmallVector<std::unique_ptr<InputFile>, 0> BitcodeCompiler::compile() {
+  uint64_t size;
+  ec = sys::fs::file_size(name, size);
+  if (ec) {
+    error("Cannot get the size of file " + name + ": " + ec.message());
```
Fixed.

smithp35 commented on lld/test/ELF/gnu-lto/hello.test:
```diff
@@ -0,0 +1,30 @@
+// RUN: llvm-readelf -s %t.pie | \
+// RUN:   FileCheck --check-prefix=READELF %s
+
+#include <stdio.h>
```
I recommend removing any dependency on the C-library. If we don't need the output to run, just that the link succeeded then 
```int f(void) { return 0; }```
Would suffice. 

This would permit the test to work with just GCC being able to compile a file, but without needing all the include and library paths configured.

smithp35 reviewed (commented):
Thanks very much for the update. My main concern in the original patch was the optional inclusion of a GPL header, which is now gone; thank you.

Will be out of office for the next 3 weeks so may not be able to respond to any further updates. I've got no fundamental objections, so happy for other reviewers to proceed without me.

tuliom committed:
Improve help text for -plugin

tuliom committed:
Remove dependency on the C-library from gnu-lto/hello.test

tuliom committed:
Move class GccIRCompiler to its own file

Remove GccIRCompiler code from LTO.h and LTO.cpp.

tuliom committed:
Remove unnecessary headers

tuliom committed:
Describe why we need global variable gcc

tuliom committed:
Move LLD_ENABLE_GNU_LTO around the code that parses GCC collect2 options

tuliom committed:
Error out when LLVM bitcode files are mixed with GCC IR files

Also add a test for this.

tuliom committed:
Mention -plugin in LLD's man page

smithp35 reviewed (commented):
Thanks for the updates. I'm happy with the resolutions. As mentioned in last comment I'm out of office for the next few weeks, hopefully other reviewers will be able to progress in my absence.

tuliom committed:
Merge branch 'main' into lld-plugin

tuliom committed:
Fix a conflict

Fix a conflict with the previous merge that was not properly fixed.

MaskRay reviewed (commented):
This needs a rebase after recent DTLTO changes.

MaskRay commented on lld/ELF/Options.td:
```diff
@@ -799,15 +799,17 @@ def: J<"plugin-opt=thinlto-prefix-replace=">,
-defm plugin: Eq<"plugin", "Ignored for compatibility with GNU linkers">;
+defm plugin
+    : Eq<"plugin", "Use a plugin in the linking process. Only supports GNU LTO "
+                   "plugin. Optional feature, may not be available.">;
```
Omit trailing `.` in help message

MaskRay commented on lld/include/lld/config.h.cmake:
```diff
@@ -0,0 +1,13 @@
+#ifndef CONFIG_H
```
Is this expected to be exported so that other projects can include it? In llvm,

* config.h: This file is an internal header used primarily within the LLVM source tree itself
* llvm-config.h: This file is the public, stable interface for external projects that want to use LLVM libraries

MaskRay commented on lld/test/ELF/plugin.test:
```diff
@@ -0,0 +1,11 @@
+RUN: not ld.lld -plugin foo  2>&1 | FileCheck --check-prefix=MISSING %s
+RUN: not ld.lld --plugin=foo 2>&1 | FileCheck --check-prefix=MISSING %s
+
+MISSING: Cannot find plugin foo
```
`MISSING: error: ....`

and use `--implicit-check-not=error:` to ensure that there is no other error.

EliasSC30 commented on lld/CMakeLists.txt:
I don't think the '!' can be used in <condition> expressions in cmake.
